### PR TITLE
Integrate tiny vec to named vectors

### PR DESCRIPTION
**Pull request by IvanPleshkov**
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [ ] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [ ] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/data_types/tiny_map.rs [hunk 1]
```diff
@@ -1,12 +1,21 @@
 use tinyvec::TinyVec;
 
-const CAPACITY: usize = 3;
-
-pub struct TinyMap<K: Default, V: Default> {
-    list: TinyVec<[(K, V); CAPACITY]>,
+pub const CAPACITY: usize = 3;
+
+#[derive(Clone, PartialEq)]
```
Comment (generall) [inline lib/segment/src/data_types/tiny_map.rs:5]: I am afraid we can't derive `PartialEq` here, underlying list will be order sensitive, which is incorrect for the map. I would propose to make custom eq by looking all keys from one man in another + checking the length

File: lib/segment/src/data_types/named_vectors.rs [hunk 2]
```diff
@@ -67,7 +70,7 @@ impl<'a> NamedVectors<'a> {
     }
 
     pub fn keys(&self) -> impl Iterator<Item = &str> {
-        self.map.keys().map(|k| k.as_ref())
+        self.map.list.iter().map(|(k, _)| k.as_ref())
```
Comment (generall) [inline lib/segment/src/data_types/named_vectors.rs:73]: why not use `.map.keys()`?

File: lib/segment/src/data_types/tiny_map.rs (hunk 1 above)
Comment (IvanPleshkov) [reply lib/segment/src/data_types/tiny_map.rs:5]: You are right, will implement trait manually

File: lib/segment/src/data_types/named_vectors.rs (hunk 2 above)
Comment (IvanPleshkov) [reply lib/segment/src/data_types/named_vectors.rs:73]: Some strange lifetime error appeared. I got unresolved lifetimes locally

Comment (generall) [reply lib/segment/src/data_types/named_vectors.rs:73]: with current approach it looks like abstraction leak

**Review by generall — APPROVED** [review]

File: lib/segment/src/data_types/named_vectors.rs [hunk 3]
```diff
@@ -1,20 +1,21 @@
 use std::borrow::Cow;
 use std::collections::HashMap;
 
+use super::tiny_map::TinyMap;
```
Comment (e-ivkov) [inline lib/segment/src/data_types/named_vectors.rs:4]:
```suggestion
use super::tiny_map;
```

File: lib/segment/src/data_types/named_vectors.rs [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
 use super::vectors::{VectorElementType, DEFAULT_VECTOR_NAME};
 
 type CowKey<'a> = Cow<'a, str>;
 type CowValue<'a> = Cow<'a, [VectorElementType]>;
-type HashMapType<'a> = HashMap<CowKey<'a>, CowValue<'a>>;
+type HashMapType<'a> = TinyMap<CowKey<'a>, CowValue<'a>>;
```
Comment (e-ivkov) [inline lib/segment/src/data_types/named_vectors.rs:9]:
```suggestion
type TinyMap<'a> = tiny_map::TinyMap<CowKey<'a>, CowValue<'a>>;
```

Comment (e-ivkov) [reply lib/segment/src/data_types/named_vectors.rs:9]: `HashMapType` just does not sound right, I assume you wanted a new typename but `HashMap` is already imported here, so this is my suggestion

Comment (IvanPleshkov) [reply lib/segment/src/data_types/named_vectors.rs:9]: Thanks for reviewing, will fix

**Pull request merged**